File: evaluation/common/src/main/java/cz/cuni/mff/java/kurinna/common/utils/QueryExecutor.java
package cz.cuni.mff.java.kurinna.common.utils;

import org.openjdk.jmh.results.RunResult;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;
import org.openjdk.jmh.util.Statistics;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

/**
 * Executes a query supplier under JMH measurement.
 *
 * Replaces the previous hand-rolled JFR implementation. JMH handles:
 * <ul>
 *   <li>One warmup iteration (discarded from stats, matching the old behaviour
 *       of the orchestrator ignoring the first run)</li>
 *   <li>GC between every iteration via {@code shouldDoGC(true)}</li>
 *   <li>Dead-code elimination prevention via {@link org.openjdk.jmh.infra.Blackhole}</li>
 *   <li>Statistical aggregation (mean / min / max)</li>
 * </ul>
 *
 * Memory is measured per-iteration inside {@link DynamicBenchmark} using
 * {@code com.sun.management.ThreadMXBean#getThreadAllocatedBytes}, which is
 * the same per-thread allocation counter the previous JFR recording relied on.
 *
 * <p><b>Concurrency note:</b> uses a static volatile field in
 * {@link SupplierHolder} — not safe if two requests benchmark the same
 * microservice concurrently. The orchestrator calls microservices sequentially
 * so this is not a problem in practice.</p>
 */
public class QueryExecutor {

    /**
     * Executes the given query supplier under JMH and returns aggregated
     * timing and memory metrics together with per-iteration detail.
     *
     * @param supplier    the query to benchmark (must return a non-null Collection)
     * @param repetitions total number of runs; the first is treated as warmup
     *                    and excluded from statistics (minimum effective value: 2)
     * @return map with keys: elapsed, minTime, maxTime, delta, minMemory,
     *         maxMemory, repetitions, iterationResults, result, status
     */
    public static <T extends Collection> Map<String, Object> executeWithMeasurement(
            Supplier<T> supplier, int repetitions) {

        SupplierHolder.current = supplier;
        SupplierHolder.iterationResults.clear();

        Map<String, Object> response = new HashMap<>();

        try {
            // this mimics the original - not a good design
            // now it is repetitions + warmup
            //int measurementIterations = Math.max(1, repetitions - 1);
            int measurementIterations = Math.max(1, repetitions );

            Options opt = new OptionsBuilder()
                    .include(DynamicBenchmark.class.getName())
                    .forks(0)                    // must be 0 — already inside a running JVM
                    .warmupIterations(1)         // first run discarded, matching old behaviour
                    .measurementIterations(measurementIterations)
                    .shouldDoGC(true)            // replaces the old System.gc() + sleep(100)
                    .build();

            Collection<RunResult> jmhResults = new Runner(opt).run();
            RunResult runResult = jmhResults.iterator().next();

            // JMH primary result covers only measurement iterations
            Statistics timeStats = runResult.getPrimaryResult().getStatistics();

            // iterationResults has warmup(1) + measurement entries; drop the warmup entry
            List<Map<String, Object>> allIters =
                    new ArrayList<>(SupplierHolder.iterationResults);
            List<Map<String, Object>> measurementIters =
                    allIters.size() > 1 ? allIters.subList(1, allIters.size()) : allIters;

            // Memory stats derived from per-iteration ThreadMXBean data
            double avgDelta = measurementIters.stream()
                    .mapToLong(m -> ((Number) m.get("delta")).longValue())
                    .average()
                    .orElse(0.0);
            long minDelta = measurementIters.stream()
                    .mapToLong(m -> ((Number) m.get("delta")).longValue())
                    .min()
                    .orElse(0L);
            long maxDelta = measurementIters.stream()
                    .mapToLong(m -> ((Number) m.get("delta")).longValue())
                    .max()
                    .orElse(0L);

            response.put("elapsed",     timeStats.getMean());
            response.put("minTime",     timeStats.getMin());
            response.put("maxTime",     timeStats.getMax());
            response.put("delta",       avgDelta);
            response.put("minMemory",   minDelta);
            response.put("maxMemory",   maxDelta);
            response.put("result",      SupplierHolder.lastSize);
            response.put("repetitions", measurementIters.size());
            response.put("iterationResults", new ArrayList<>(measurementIters));
            response.put("status",      "success");

        } catch (Exception e) {
            response.put("status", "error");
            response.put("error",  e.getMessage());
        } finally {
            SupplierHolder.current = null;
            SupplierHolder.iterationResults.clear();
        }

        return response;
    }
}

File: evaluation/common-csharp/Utils/QueryExecutor.cs
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Engines;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Loggers;
using BenchmarkDotNet.Reports;
using BenchmarkDotNet.Running;
using BenchmarkDotNet.Toolchains.InProcess.NoEmit;

namespace CommonCSharp.Utils
{
    /// <summary>
    /// Executes a query function under BenchmarkDotNet measurement.
    ///
    /// Mirrors Java's QueryExecutor + DynamicBenchmark + SupplierHolder pattern.
    /// BenchmarkDotNet handles:
    ///   - One warmup iteration (discarded from stats, matching JMH warmupIterations=1)
    ///   - GC before every iteration via DynamicBenchmark.[IterationSetup] (matching shouldDoGC(true))
    ///   - Dead-code elimination prevention by consuming the return value
    ///   - Statistical aggregation (mean / min / max)
    ///
    /// InProcessNoEmitToolchain is the .NET equivalent of JMH's forks(0):
    /// the benchmark runs inside the already-running ASP.NET process.
    ///
    /// RunStrategy.Monitoring with InvocationCount=1 is the equivalent of JMH's
    /// Mode.SingleShotTime: each iteration invokes the benchmark exactly once, so
    /// every recorded IterationResult corresponds to a single query execution.
    /// Monitoring is used rather than ColdStart because ColdStart silently ignores
    /// WithWarmupCount (it performs no warmup), which would leave the first measured
    /// iteration cold and un-JITted; Monitoring keeps the one-invocation-per-iteration
    /// semantics while honouring the warmup iteration.
    ///
    /// Because the warmup iteration runs through DynamicBenchmark, it is recorded in
    /// BenchmarkHolder.IterationResults as the first entry and is stripped below
    /// (mirroring how the Java side discards its JMH warmup).
    ///
    /// Memory is measured per-iteration inside DynamicBenchmark using
    /// GC.GetAllocatedBytesForCurrentThread(), the .NET equivalent of
    /// ThreadMXBean.getThreadAllocatedBytes().
    ///
    /// Concurrency note: uses static state in BenchmarkHolder — not safe if two
    /// requests benchmark concurrently. The orchestrator calls microservices
    /// sequentially so this is not a problem in practice.
    /// </summary>
    public static class QueryExecutor
    {
        public static Task<QueryResult> ExecuteWithMeasurement<T>(Func<Task<List<T>>> queryFn, int repetitions)
        {
            return ExecuteWithMeasurement(async () => (await queryFn()).Count, repetitions);
        }

        public static Task<QueryResult> ExecuteWithMeasurement(Func<Task<int>> queryFn, int repetitions)
        {
            BenchmarkHolder.Current = queryFn;
            lock (BenchmarkHolder.IterationResults)
                BenchmarkHolder.IterationResults.Clear();

            try
            {
                int measurementIterations = Math.Max(1, repetitions);

                var config = ManualConfig.CreateEmpty()
                    .WithOptions(ConfigOptions.DisableOptimizationsValidator)
                    .AddJob(Job.Default
                        .WithToolchain(InProcessNoEmitToolchain.Instance)
                        .WithStrategy(RunStrategy.Monitoring)
                        .WithInvocationCount(1)   // one query per iteration (SingleShotTime)
                        .WithWarmupCount(1)       // one discarded warmup iteration (JMH warmupIterations=1)
                        .WithIterationCount(measurementIterations))
                    .AddLogger(NullLogger.Instance);

                Summary summary = BenchmarkRunner.Run<DynamicBenchmark>(config);
                BenchmarkReport report = summary.Reports.First();

                // BDN timing for measurement iterations only (warmup excluded)
                var bdnTimingsMs = report.AllMeasurements
                    .Where(m => m.Is(IterationMode.Workload, IterationStage.Actual))
                    .Select(m => m.Nanoseconds / 1_000_000.0)
                    .ToList();

                // Per-iteration results captured by DynamicBenchmark. Under Monitoring
                // the warmup iteration runs through DynamicBenchmark and is recorded as
                // the first entry, so strip it (mirrors the Java side discarding the
                // JMH warmup). The remaining entries are the real measurement iterations.
                List<IterationResult> allIters;
                lock (BenchmarkHolder.IterationResults)
                    allIters = new List<IterationResult>(BenchmarkHolder.IterationResults);

                var measurementIters = allIters.Count > 1
                    ? allIters.Skip(1).ToList()
                    : allIters;

                var result = new QueryResult
                {
                    elapsed   = bdnTimingsMs.Count > 0 ? bdnTimingsMs.Average() : 0,
                    minTime   = bdnTimingsMs.Count > 0 ? bdnTimingsMs.Min()     : 0,
                    maxTime   = bdnTimingsMs.Count > 0 ? bdnTimingsMs.Max()     : 0,
                    delta     = measurementIters.Count > 0 ? measurementIters.Average(r => (double)r.delta) : 0,
                    minMemory = measurementIters.Count > 0 ? (double)measurementIters.Min(r => r.delta)     : 0,
                    maxMemory = measurementIters.Count > 0 ? (double)measurementIters.Max(r => r.delta)     : 0,
                    result       = BenchmarkHolder.LastSize,
                    repetitions  = measurementIters.Count,
                    iterationResults = measurementIters,
                    status = "success"
                };

                return Task.FromResult(result);
            }
            catch (Exception ex)
            {
                return Task.FromResult(new QueryResult { status = "error", error = ex.Message });
            }
            finally
            {
                BenchmarkHolder.Current = null;
                lock (BenchmarkHolder.IterationResults)
                    BenchmarkHolder.IterationResults.Clear();
            }
        }
    }
}
